src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -700,32 +700,71 @@ public void ExecuteCodeBlockActions<TLanguageKindEnum>(
+
+                    var codeBlockScope = new HostCodeBlockStartAnalysisScope<TLanguageKindEnum>(startAction.Analyzer);
+                    var blockStartContext = new AnalyzerCodeBlockStartAnalysisContext<TLanguageKindEnum>(
+                        codeBlockScope, declaredNode, declaredSymbol, semanticModel, @this.AnalyzerOptions, filterSpan, isGeneratedCode, cancellationToken);
+
+                    // Catch Exception from the start action.
+                    @this.ExecuteAndCatchIfThrows(
+                        startAction.Analyzer,
+                        static data =>
+                        {
+                            data.action(data.context);
+                            data.blockEndActions?.AddAll(data.scope.CodeBlockEndActions);
+                            data.syntaxNodeActions?.AddRange(data.scope.SyntaxNodeActions);
+                        },
+                        (action: startAction.Action, context: blockStartContext, scope: codeBlockScope, blockEndActions: codeBlockEndActions, syntaxNodeActions),
+                        new AnalysisContextInfo(@this.Compilation, declaredSymbol, declaredNode),
+                        cancellationToken);
+                },
+                executeActions: static (syntaxNodeActions, diagReporter, isSupportedDiagnostic, args, cancellationToken) =>
+                {
+                    var (@this, codeBlockStartActions, analyzer, declaredNode, declaredSymbol, executableCodeBlocks, semanticModel, getKind, filterSpan, isGeneratedCode) = args;
+
+                    Debug.Assert(getKind != null);
+
+                    var executableNodeActionsByKind = GetNodeActionsByKind(syntaxNodeActions);
+                    var syntaxNodesToAnalyze = executableCodeBlocks.SelectMany(cb =>
                     {
                         var filter = semanticModel.GetSyntaxNodesToAnalyzeFilter(cb, declaredSymbol);
 
-                        if (filter is object)
-                        {
-                            return cb.DescendantNodesAndSelf(descendIntoChildren: filter).Where(filter);
-                        }
-                        else
-                        {
-                            return cb.DescendantNodesAndSelf();
-                        }
-                    }),
-                semanticModel, getKind, filterSpan, isGeneratedCode, cancellationToken);
+                        return filter is not null
+                            ? cb.DescendantNodesAndSelf(descendIntoChildren: filter).Where(filter)
+                            : cb.DescendantNodesAndSelf();
+                    });
+
+                    @this.ExecuteSyntaxNodeActions(syntaxNodesToAnalyze, executableNodeActionsByKind, analyzer, declaredSymbol, semanticModel, getKind, diagReporter, isSupportedDiagnostic, filterSpan, isGeneratedCode, hasCodeBlockStartOrSymbolStartActions: codeBlockStartActions.Any(), cancellationToken);
```

Comment:
>syntaxNodesToAnalyze [](http://example.com/codeflow?start=51&length=20)

Any thoughts on moving this one away from IEnumerable too?

Reply:
that's a goal i have for a future PR. for this one i don't want to change the *actual* types that we are using today.  Today this *is* a lazy linq enumerable, so i'm keeping that.  It would make sense for thsi to instead use a scratch buffer and pool that in a followup PR.

Reply:
decided to just do this here.  a pooled builder should really be better than linq.